tests/map_test_case.h

```diff
@@ -0,0 +1,341 @@
+                                           std::function<void( tile, std::ostringstream &out )> &callback );
+
+        /**
+         * Invokes callback for each tile of the input map.
+         * Collects results of the invocation in 2d vector.
+         */
+        template<typename R>
+        std::vector<std::vector<R>> map_tiles( const std::function<R( tile )> &callback );
+
+        // input major (first) dimension
+        int get_height() const;
+
+        // input minor dimension
+        int get_width() const;
+
+        // returns calculated "origin" point, i.e. shift relative to the map
+        tripoint get_origin();
+
+        // automatically set `anchor_char` from the list of given chars to the one that is present in the `setup`
+        void set_anchor_char_from( const std::set<char> &chars );
+
+        // WARNING: transformations might change later results of getters (`get_origin`, `get_height`, etc.)
+
+        // transposes `setup` and `expected_results`
+        void transpose();
+
+        // reverses rows in `setup` and `expected_results`
+        void reflect_x();
+
+        // reverses columns in `setup` and `expected_results`
+        void reflect_y();
+
+        /**
+        * Generates (using catch2 GENERATE) every combination of three transformations and applies them to the test.
+        * See https://github.com/catchorg/Catch2/blob/master/docs/generators.md
+        * Test will be invoked for each combination of transformation.
+        * @param t test case to apply transformations to
+        * @return string info about applied transformations
+        */
+        std::string generate_transform_combinations();
+
+        /**
+         * Ensures that `map_test_case` is set up correctly by
+         * validating given point in map coordinates against `map_test_case::anchor_map_pos`
+         * (and performing other sanity checks)
+         * @param p tripoint in map coordinates that is expected to align with the `anchor_map_pos`
+         */
+        void validate_anchor_point( tripoint p );
+
+    private:
+        // origin (0,0) of this `map_test_case` in `map` coordinates
+        // based on `anchor_map_pos` and `anchor_char`, lazily calculated when needed, reset on transformations
+        cata::optional<tripoint> origin = cata::nullopt;
+
+        // flag that internal sanity checks are completed, resets on transformations
+        bool checks_complete = false;
+
+        void do_internal_checks();
+
+        void for_each_tile( tripoint tmp_origin, const std::function<void( tile & )> &callback ) const;
+
+};
+
+// common helpers, used together with map_test_case
+namespace map_test_case_common
+{
+/**
+ * Combines two arbitrary functions `f` and `g` into a single one.
+ * `f` and `g` must have same signature
+ */
+template<typename R, typename ... Args>
+static inline std::function<R( Args ... )> operator+(
+    const std::function<R( Args ... )> f,
+    const std::function<R( Args ... )> g )
```

Comment:
It's not generally a good idea to overload operators for types that are not under your control, and especially not for types from the standard library.  In Rust, this is called *coherence*, and is enforced by the compiler.  In C++ it's merely good practice.

Because you're writing these in a special namespace that's specific to this header you can probably get away with it in practice, but if you wanted to fix it there are two obvious ways:
* Make these named functions, rather than operators, or
* Make a trivial wrapper class around `std::function`, and overload the operators for that class instead.
Alternatively, you could use a small number of more focused types that better capture the semantics here, like `tile_predicate` for `bool(tile)` functions.  But that might lead to more code duplication.

Reply:
I limited the overloaded operators to the `tile_predicate` type. That actually helped to reduce boilerplate a bit.